deps/rabbit/src/rabbit_global_counters.erl

```diff
@@ -0,0 +1,183 @@
+            ]).
+
+%% Protocol & QueueType counters:
+-define(MESSAGES_DELIVERED, 1).
+-define(MESSAGES_DELIVERED_CONSUME_MANUAL_ACK, 2).
+-define(MESSAGES_DELIVERED_CONSUME_AUTO_ACK, 3).
+-define(MESSAGES_DELIVERED_GET_MANUAL_ACK, 4).
+-define(MESSAGES_DELIVERED_GET_AUTO_ACK, 5).
+-define(MESSAGES_GET_EMPTY, 6).
+-define(MESSAGES_REDELIVERED, 7).
+-define(MESSAGES_ACKNOWLEDGED, 8).
+-define(PROTOCOL_QUEUE_TYPE_COUNTERS,
+            [
+                {
+                    messages_delivered_total, ?MESSAGES_DELIVERED, counter,
+                    "Total number of messages delivered to consumers"
+                },
+                {
+                    messages_delivered_consume_manual_ack_total, ?MESSAGES_DELIVERED_CONSUME_MANUAL_ACK, counter,
+                    "Total number of messages delivered to consumers using basic.consume with manual acknowledgment"
+                },
+                {
+                    messages_delivered_consume_auto_ack_total, ?MESSAGES_DELIVERED_CONSUME_AUTO_ACK, counter,
+                    "Total number of messages delivered to consumers using basic.consume with automatic acknowledgment"
+                },
+                {
+                    messages_delivered_get_manual_ack_total, ?MESSAGES_DELIVERED_GET_MANUAL_ACK, counter,
+                    "Total number of messages delivered to consumers using basic.get with manual acknowledgment"
+                },
+                {
+                    messages_delivered_get_auto_ack_total, ?MESSAGES_DELIVERED_GET_AUTO_ACK, counter,
+                    "Total number of messages delivered to consumers using basic.get with automatic acknowledgment"
+                },
+                {
+                    messages_get_empty_total, ?MESSAGES_GET_EMPTY, counter,
+                    "Total number of times basic.get operations fetched no message"
+                },
+                {
+                    messages_redelivered_total, ?MESSAGES_REDELIVERED, counter,
+                    "Total number of messages redelivered to consumers"
+                },
+                {
+                    messages_acknowledged_total, ?MESSAGES_ACKNOWLEDGED, counter,
+                    "Total number of messages acknowledged by consumers"
+                }
+            ]).
+
+boot_step() ->
+    init([{protocol, amqp091}]),
+    init([{protocol, amqp091}, {queue_type, rabbit_classic_queue}]),
+    init([{protocol, amqp091}, {queue_type, rabbit_quorum_queue}]),
+    init([{protocol, amqp091}, {queue_type, rabbit_stream_queue}]).
+
+init(Labels = [{protocol, Protocol}, {queue_type, QueueType}]) ->
+    _ = seshat_counters:new_group(?MODULE),
+    Counters = seshat_counters:new(?MODULE, Labels, ?PROTOCOL_QUEUE_TYPE_COUNTERS),
+    persistent_term:put({?MODULE, Protocol, QueueType}, Counters),
+    ok;
+init(Labels = [{protocol, Protocol}]) ->
+    _ = seshat_counters:new_group(?MODULE),
+    Counters = seshat_counters:new(?MODULE, Labels, ?PROTOCOL_COUNTERS),
```

Comment:
You are right, they don't, but the way seshat currently works would require us to duplicate counters. We picked up on this issue, but the simplest alternative makes things worse. For example, this is what we have now:

```erlang
%% PROTOCOL COUNTERS:
-define(MESSAGES_RECEIVED, 1).
-define(MESSAGES_RECEIVED_CONFIRM, 2).
-define(MESSAGES_ROUTED, 3).
-define(MESSAGES_UNROUTABLE_DROPPED, 4).
-define(MESSAGES_UNROUTABLE_RETURNED, 5).
-define(MESSAGES_CONFIRMED, 6).
-define(PROTOCOL_COUNTERS,
            [
                {
                    messages_received_total, ?MESSAGES_RECEIVED, counter,
                    "Total number of messages received from publishers"
                },
                {
                    messages_received_confirm_total, ?MESSAGES_RECEIVED_CONFIRM, counter,
                    "Total number of messages received from publishers expecting confirmations"
                },
                {
                    messages_routed_total, ?MESSAGES_ROUTED, counter,
                    "Total number of messages routed to queues or streams"
                },
                {
                    messages_unroutable_dropped_total, ?MESSAGES_UNROUTABLE_DROPPED, counter,
                    "Total number of messages published as non-mandatory into an exchange and dropped as unroutable"
                },
                {
                    messages_unroutable_returned_total, ?MESSAGES_UNROUTABLE_RETURNED, counter,
                   "Total number of messages published as mandatory into an exchange and returned to the publisher as unroutable"
                },
                {
                    messages_confirmed_total, ?MESSAGES_CONFIRMED, counter,
                    "Total number of messages confirmed to publishers"
                }
            ]).
```

And this is the simplest change that we could make, which doesn't look like a step in the right direction (more code is required, but this is the bulk of it):

```erlang
%% STREAM PROTOCOL COUNTERS:
-define(STREAM_MESSAGES_RECEIVED, 1).
-define(STREAM_MESSAGES_RECEIVED_CONFIRM, 2).
-define(STREAM_MESSAGES_CONFIRMED, 3).
-define(STREAM_PROTOCOL_COUNTERS,
            [
                {
                    messages_received_total, ?STREAM_MESSAGES_RECEIVED, counter,
                    "Total number of messages received from publishers"
                },
                {
                    messages_received_confirm_total, ?STREAM_MESSAGES_RECEIVED_CONFIRM, counter,
                    "Total number of messages received from publishers expecting confirmations"
                },
                {
                    messages_confirmed_total, ?STREAM_MESSAGES_CONFIRMED, counter,
                    "Total number of messages confirmed to publishers"
                }
            ]).

%% AMQP091 PROTOCOL COUNTERS:
-define(AMQP091_MESSAGES_RECEIVED, 1).
-define(AMQP091_MESSAGES_RECEIVED_CONFIRM, 2).
-define(AMQP091_MESSAGES_ROUTED, 3).
-define(AMQP091_MESSAGES_UNROUTABLE_DROPPED, 4).
-define(AMQP091_MESSAGES_UNROUTABLE_RETURNED, 5).
-define(AMQP091_MESSAGES_CONFIRMED, 6).
-define(AMQP091_PROTOCOL_COUNTERS,
            [
                {
                    messages_received_total, ?AMQP091_MESSAGES_RECEIVED, counter,
                    "Total number of messages received from publishers"
                },
                {
                    messages_received_confirm_total, ?AMQP091_MESSAGES_RECEIVED_CONFIRM, counter,
                    "Total number of messages received from publishers expecting confirmations"
                },
                {
                    messages_routed_total, ?AMQP091_MESSAGES_ROUTED, counter,
                    "Total number of messages routed to queues or streams"
                },
                {
                    messages_unroutable_dropped_total, ?AMQP091_MESSAGES_UNROUTABLE_DROPPED, counter,
                    "Total number of messages published as non-mandatory into an exchange and dropped as unroutable"
                },
                {
                    messages_unroutable_returned_total, ?AMQP091_MESSAGES_UNROUTABLE_RETURNED, counter,
                   "Total number of messages published as mandatory into an exchange and returned to the publisher as unroutable"
                },
                {
                    messages_confirmed_total, ?AMQP091_MESSAGES_CONFIRMED, counter,
                    "Total number of messages confirmed to publishers"
                }
            ]).
```

Rather than doing the above, we (+@dumbbell) want to go for a refactor in seshat, which will allow counters to be initialised independently, without hard-coding the context (e.g. protocol type) into the counters themselves.

Another good spot @ansd which confirms that what we picked up on bugs you too 😉

Reply:
Please re-open @ansd if you feel this is not resolved 🧐